src/TerrainTile.cc

```diff
@@ -31,140 +31,128 @@ const char*  TerrainTile::_jsonMinElevationKey  = "min";
+    _data = new int16_t*[_tileInfo.gridSizeLat];
+    for (int k = 0; k < _tileInfo.gridSizeLat; k++) {
+        _data[k] = new int16_t[_tileInfo.gridSizeLon];
     }
 
     int valueIndex = 0;
     const int16_t* pTileData = reinterpret_cast<const int16_t*>(&reinterpret_cast<const uint8_t*>(byteArray.constData())[cTileHeaderBytes]);
-    for (int i = 0; i < _gridSizeLat; i++) {
-        for (int j = 0; j < _gridSizeLon; j++) {
+    for (int i = 0; i < _tileInfo.gridSizeLat; i++) {
+        for (int j = 0; j < _tileInfo.gridSizeLon; j++) {
             _data[i][j] = pTileData[valueIndex++];
         }
     }
 
     _isValid = true;
+}
 
-    return;
+TerrainTile::~TerrainTile()
+{
+    if (!_data) {
+        return;
+    }
+
+    for (unsigned i = 0; i < static_cast<unsigned>(_tileInfo.gridSizeLat); i++) {
+        delete[] _data[i];
+    }
+
+    delete[] _data;
 }
 
 double TerrainTile::elevation(const QGeoCoordinate& coordinate) const
 {
-    if (_isValid && _southWest.isValid() && _northEast.isValid()) {
-        qCDebug(TerrainTileLog) << "elevation: " << coordinate << " , in sw " << _southWest << " , ne " << _northEast;
-
-        // The lat/lon values in _northEast and _southWest coordinates can have rounding errors such that the coordinate
-        // request may be slightly outside the tile box specified by these values. So we clamp the incoming values to the
-        // edges of the tile if needed.
-
-        double clampedLon = qMax(coordinate.longitude(), _southWest.longitude());
-        double clampedLat = qMax(coordinate.latitude(), _southWest.latitude());
-
-        // Calc the index of the southernmost and westernmost index data value
-        int lonIndex = qFloor((clampedLon - _southWest.longitude()) / tileValueSpacingDegrees);
-        int latIndex = qFloor((clampedLat - _southWest.latitude()) / tileValueSpacingDegrees);
-
-        // Calc how far along in between the known values the requested lat/lon is fractionally
-        double lonIndexLongitude    = _southWest.longitude() + (static_cast<double>(lonIndex) * tileValueSpacingDegrees);
-        double lonFraction          = (clampedLon - lonIndexLongitude) / tileValueSpacingDegrees;
-        double latIndexLatitude     = _southWest.latitude() + (static_cast<double>(latIndex) * tileValueSpacingDegrees);
-        double latFraction          = (clampedLat - latIndexLatitude) / tileValueSpacingDegrees;
-
-        // Calc the elevation as the average across the four known points
-        double known00      = _data[latIndex][lonIndex];
-        double known01      = _data[latIndex][lonIndex+1];
-        double known10      = _data[latIndex+1][lonIndex];
-        double known11      = _data[latIndex+1][lonIndex+1];
-        double lonValue1    = known00 + ((known01 - known00) * lonFraction);
-        double lonValue2    = known10 + ((known11 - known10) * lonFraction);
-        double latValue     = lonValue1 + ((lonValue2 - lonValue1) * latFraction);
```

Comment:
Dropped the bi-linear interpolation since there is no real benefit in doing so. Simple interpolation methods are not suited for terrain data. More advanced interpolation algorithms based on statistical methods, e.g. Kriging, would be ideal. Nearest neighbor and linear interpolation yield about the same error on average.